Added project files.

Larisa-Antsifrova opened the pull request:
Done:
- Added files from the case task.

Remember:
- Deadline 6.08.2021 23:59.

Happy reviewing :)

MazanYan commented on repositories/users-repository.js:
```diff
@@ -0,0 +1,45 @@
+    const allUsers = await getAllUsers(usersPath);
+    allUsers.push(newUser);
+
+    await fs.writeFile(usersPath, JSON.stringify(allUsers, null, 2));
```
Here (in lines 34-37) we are reading the entire file, adding a new user, and completely rewriting the file. It's not effective in terms of RAM usage and time effeciency, especially when we will have a lot of users registered.

Larisa-Antsifrova replied:
I guess further actions on this one would depend on your suggestion to the thread below :)

MazanYan commented on db/users.json:
```diff
@@ -0,0 +1,8 @@
+[
```
Here we are basically commiting a file with personal data and making it public. Shouldn't it be added to `.gitignore`? And will the application recreate the file if it gets deleted?

Larisa-Antsifrova replied:
That's true :)
In the scope of a test task, I left the DB folder and its file to stay in the project structure. 
So far, no. The DB file is expected to exist)

MazanYan commented on index.js:
```diff
@@ -0,0 +1,44 @@
+  });
+});
+
+process.on('unhandledRejection', (reason, promise) => {
```
I didn't know how to handle this promise rejection error before. It was interesting to know.

MazanYan commented on helpers/is-unique.js:
```diff
@@ -0,0 +1,17 @@
+const isUniqueUser = async email => {
+  try {
+    const allUsers = await getAllUsers(usersPath);
+    const isUnique = allUsers.find(user => user.email === email);
```
This operation has a linear time complexity what is not effective. I'd suggest to use the Front-end normalization on the allUsers variable and data structure.
Changing from this:
`const allUsers = [
  {
    email: 'kakashi-hatake@mail.com',
    ...
  },
  ...
];`

to this: 
`const allUsers = {
  'kakashi-hatake@mail.com': {
    ...
  },
  'other-user@mail.com': {
    ...
  },
  ...
};`

Then we can find the user using `allUsers[email]` only.

Larisa-Antsifrova replied:
Nice, that's the first time I've heard of Front-end normalization (I've read this article and I guess that's what you meant: https://alfredayibonte.medium.com/front-end-normalization-8e3d0df7cb47). So, I will ask before action:
Do you suggest it's more efficient to change the structure of data stored in the "database" file from an array of objects to one big object containing all the users by their e-mails as their keys?

MazanYan replied:
Yes, I meant this normalization. I completed this task by storing all users in the corresponding directory. The `users` directory contained the separate files with users' e-mails as names (user entity primary key). I'm not sure if `fs.readFile(name, callback)` finds the necessary file in constant time but I'm sure it's a more straighforward approach.

Larisa-Antsifrova replied:
Aha, okay, I see, so that also helped to avoid re-writing the whole file.

MazanYan commented on package.json:
```diff
@@ -0,0 +1,27 @@
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/Larisa-Antsifrova/project-ses-bit-web-api.git"
```
The repository URL is different now

Larisa-Antsifrova replied:
Haha! V - vnimatelnost. Thank you, Yan! Fixed that :)

MazanYan commented on package.json:
```diff
@@ -0,0 +1,27 @@
+  "description": "Test task for SES.",
+  "main": "index.js",
+  "scripts": {
+    "start:dev": "nodemon index.js"
```
Will we have `start:debug` or `start:prod` scripts? If not, we can rename it to `start` only

Larisa-Antsifrova replied:
I got your point! I have added the _start_ script separately, the one that will be used for deployment, and left _start:dev_ for development.

MazanYan commented on services/coinlayer-api.js:
```diff
@@ -0,0 +1,27 @@
+require('dotenv').config();
+const axios = require('axios');
+
+const COINLAYER_API_KEY = process.env.COINLAYER_API_KEY;
```
Why can't we use `const { COINLAYER_API_KEY } = process.env;`? It takes less code space and does not lower readability.

Larisa-Antsifrova replied:
Good point! Thanks :) Destructured the variables in other similar places in code as well.

MazanYan commented on controllers/users-controllers.js:
```diff
@@ -0,0 +1,69 @@
+const isUniqueUser = require('../helpers/is-unique');
+const { HttpCodes } = require('../helpers/constants');
+
+const createUser = async (req, res, next) => {
```
The controller does too much. In MVC, for example the controller only transfers the user input to model (any UserService etc).

Larisa-Antsifrova replied:
Alright, this one took me a while. I've added UserService. But it does not feel to improve anything in particular. If before, the controllers dealt with all the services, now UserService depends on Password and TokenService. On the other hand, I would not like to make my UserService return status codes and response message.
I might not see the long-term benefit here. I would appreciate it, if you could explain or show example :)

MazanYan commented on controllers/users-controllers.js:
```diff
@@ -0,0 +1,69 @@
+        .json({ message: 'This email is already in use.' });
+    }
+
+    const hashedPassword = await bcrypt.hash(password, 8);
```
Why isn't password encryption a separate service?

Hellrasier replied:
Password encryption is better to go to middlewares section

Larisa-Antsifrova replied:
That's a great question. Thank you! Moved password hashing and comparing to a separate service in the next commits.

MazanYan commented on controllers/users-controllers.js:
```diff
@@ -0,0 +1,69 @@
+
+    const payload = { id: user.id, name: user.name };
+    const JWT_SECRET_KEY = process.env.JWT_SECRET_KEY;
+    const token = jwt.sign(payload, JWT_SECRET_KEY, { expiresIn: '4h' }); // access token is given for 4 hours
```
`expiresIn` clearly explains that the token expires in 4 hours. I think the comment is redundant.

Larisa-Antsifrova replied:
Noted :) The comment is deleted and the whole JWT generation is moved to a separate service as well.

MazanYan commented on controllers/users-controllers.js:
```diff
@@ -0,0 +1,69 @@
+        .json({ message: 'Invalid credentials.' });
+    }
+
+    const isPasswordCorrect = await bcrypt.compare(password, user.password);
```
I suggest to move password encryption and comparing to the separate service.

Larisa-Antsifrova replied:
Done :)

MazanYan commented on .example.env:
```diff
@@ -0,0 +1,5 @@
+
+JWT_SECRET_KEY={custom secret key for JWT sign}
+
+COINLAYER_API_KEY={api key for bitcoin rates api}
```
EOF line is missing.

Larisa-Antsifrova replied:
I get it that EOF stands for End Of File. But I am not sure I get what exactly is expected. Could you please hint me with an example? :)

MazanYan replied:
It's usual for files with code to end with a blank line. I think it serves the same way as a comma in the vertical array/function arguments like this:
```
const longArray = [
  '1',
  '2,235',
  '-1F',
];
```

It makes adding new records and lines of code more easy and makes the file structure more consistent (like with comma example - all records have a comma after them). Linters usually have this rule warning the programmer about missing line in the end of file. GitHub also shows the warning in the pull request.

Larisa-Antsifrova replied:
Wow :) Yan, thanks! That's a  new insight for me :)

>  Linters usually have this rule warning the programmer about missing line in the end of file. GitHub also shows the warning in the pull request.

Larisa-Antsifrova replied:
Done )

MazanYan reviewed (changes requested):
Thank you for a good pull request! I have a few suggestions in comments.

Hellrasier commented on controllers/users-controllers.js:
```diff
@@ -0,0 +1,69 @@
+  try {
+    const { email, password } = req.body;
+
+    const user = await User.getUserByEmail(email);
```
Storage is named as User, maybe it is better to create first more abstract "storage" than specify it only for user

Hellrasier commented on helpers/api-limiter-config.js:
```diff
@@ -0,0 +1,13 @@
+    });
+  },
+};
+
```
anti-spam and anti-dos system, gj

Hellrasier commented on services/coinlayer-api.js:
```diff
@@ -0,0 +1,27 @@
+    return { timestamp, target, rates };
+  } catch (error) {
+    console.log('Error in fetchUahToBtcRate: ', error.message);
+  }
```
Function returns undefined on error, maybe return error to properly handle it?

Larisa-Antsifrova commented:
Hi!
Thank you for the suggestions! I've done my best to address most of them:
- Moved password manipulations and token generation to their separate services
- Destructured env variables where needed
- Edited package.json file
- Removed comments
- Created separate UserService to remove some of the logic from controllers

I have resolved the comments that covered the issues I refactored (I hope that's okay). And there are mainly 3 issues left:
- EOF
- Structure of the data stored in DB file
- Controllers taking too much responsibility :)
